### PR TITLE
CLIENT-SPECIFICATION (2.0): document removal of master branch, addition of support for individual translation archives

### PR DESCRIPTION
**Pull request by kbdharun**
As promised in the chatroom, I plan to create a release informing that the `master` branch is deprecated and introduce the newer client specification after merging this PR.

Note: I have pre-referenced the tag link, which will be available once it is released. The date is tentative too, we can update it later to reflect the correct date.

## Changes

- Fix typos and grammatical errors.
- Bump version to `v2.0`, minor changes in the changelog.
- Document removal of the master branch and the added support for individual translation archives.

### CODE REVIEW

**Review by pixelcmtd — COMMENTED** [review]
I don't think we need to fight over who has the worse English, but all wording/grammar/... changes seem debatable at best to me.
And what exactly is your rationale for going to 2.0? I kinda get that we might want to make a new client spec release with a few changes having accumulated, but why would it be a major? Because it's breaking for some clients? Some changes in the previous minors were possibly breaking too

**Comment by kbdharun** [comment]
The main motive behind the change is that the `master` branch got removed, not all tldr clients are listed in wiki, there are extension, plugins which still use `master` in order to make a tagged release, I have updated the client specification which has additions too like the long options one.

File: CLIENT-SPECIFICATION.md [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. For a list of previous versions of the specification, see the [changelog section](#Changelog) below.
+This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. For the list of previous versions of this specification, see the [changelog section](#Changelog) below.
 
 The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
 
 
 ## Terminology
 
-In order to aid the understanding of this specification document, a number of terms will be defined in this section.
+Several terms will be defined in this section to aid the understanding of this specification document.
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:12]: How about this?

```suggestion
This section defines key terms that are relevant for understanding this specification document.
```

File: CLIENT-SPECIFICATION.md [hunk 2]
```diff
@@ -1,15 +1,15 @@
 # tldr-pages client specification
 
-**Current Specification Version:** 1.5
+**Current Specification Version:** 2.0
 
-This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. For a list of previous versions of the specification, see the [changelog section](#Changelog) below.
+This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. For the list of previous versions of this specification, see the [changelog section](#Changelog) below.
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:5]: IMO there's not much benefit to this change. Both versions strike me as correct, and I do have a slight preference for the first one. If you don't feel strongly about this change, my recommendation would be to revert it.

File: CLIENT-SPECIFICATION.md [hunk 3]
```diff
@@ -31,23 +31,23 @@ This section describes the standardised command-line interface (CLI) for clients
 
 ### Arguments
 
-A number of command-line options MUST be supported (unless otherwise specified) if a CLI is implemented:
+Several command-line options MUST be supported (unless otherwise specified) if a CLI is implemented:
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:34]: Let's make this explicit:


```suggestion
The following command-line options MUST be supported (unless otherwise specified) if a CLI is implemented:
```

File: CLIENT-SPECIFICATION.md [hunk 4]
```diff
@@ -31,23 +31,23 @@ This section describes the standardised command-line interface (CLI) for clients
...
 
 Option             | Required?   | Meaning
 -------------------|-------------|----------
 `-v`, `--version`  | Yes         | Shows the current version of the client, and the version of this specification that it implements.
 `-p`, `--platform` | Yes         | Specifies the platform to be used to perform the action (either listing or searching) as an argument. If this option is specified, the selected platform MUST be checked first instead of the current platform as described below.
-`-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if cache is supported.
+`-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if the cache is supported.
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:40]: Here if either go with "if a cache is supported" or "of caching is supported".

File: CLIENT-SPECIFICATION.md [hunk 5]
```diff
@@ -31,23 +31,23 @@ This section describes the standardised command-line interface (CLI) for clients
...
 `-l`, `--list`     | No          | Lists all the pages in the current platform to the standard output.
 `-L`, `--language` | No          | Specifies the preferred language for the page returned. Overrides other language detection mechanisms. See the [language section](#language) for more information.
 
-Clients MUST implement both the short and long version of an option.
+Clients MUST implement both the short and long versions of an option.
 
-Additional decoration MAY be printed if the standard output is a [TTY](http://www.linusakesson.net/programming/tty/index.php). If not, then the output MUST not contain any additional decorations. For example a page list MUST be formatted with 1 page name per line (to enable easy manipulation using standard CLI tools such as `grep` etc.).
+Additional decoration MAY be printed if the standard output is a [TTY](http://www.linusakesson.net/programming/tty/index.php). If not, then the output MUST not contain any additional decorations. For example, a page list MUST be formatted with a 1-page name per line (to enable easy manipulation using standard CLI tools such as `grep` etc.).
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:46]: I agree with the comma, but "1-page name" is incorrect, as it implies a name that occupies one page (whatever that would be 😅). What we want to say here is to have "one page-name" (abusing the hyphen to make my point across). Here's what I'd suggest:


```suggestion
Additional decoration MAY be printed if the standard output is a [TTY](http://www.linusakesson.net/programming/tty/index.php). If not, then the output MUST not contain any additional decorations. For example, a page list MUST be formatted with one page name per line (to enable easy manipulation using standard CLI tools such as `grep` etc.).
```

File: CLIENT-SPECIFICATION.md [hunk 6]
```diff
@@ -151,13 +151,13 @@ It is possible that due to this page resolution logic, the client may show a pag
 
 #### If a page is not found
 
-If a page cannot be found in _any_ platform, then it is RECOMMENDED that clients display an error message with a link to create a new issue against the `tldr-pages/tldr` GitHub repository. Said link might take the following form:
+If a page cannot be found in _any_ platform, then it is RECOMMENDED that clients display an error message with a link to create a new issue against the `tldr-pages/tldr` GitHub repository. The said link might take the following form:
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:154]: The previous form is correct and AFAIK more common. If the form doesn't appeal to you, perhaps something different like "Such a link" would be preferable? (For the record, I'm fine with the existing wording.)

File: CLIENT-SPECIFICATION.md [hunk 7]
```diff
@@ -166,11 +166,11 @@ If multiple versions of a page were found for different platforms, then a client
...
 
-The [`LANG` environment variable](https://www.gnu.org/software/gettext/manual/html_node/Locale-Environment-Variables.html) specifies the user preferred locale (in the form `ll[_CC][.encoding]`). The [`LANGUAGE` environment variable](https://www.gnu.org/software/gettext/manual/html_node/The-LANGUAGE-variable.html) specifies a priority list of locales (in the form `l1:l2:...`) that can be used if the locale defined by `LANG` is not available. Both `LANG` and `LANGUAGE` may contain the values `C` or `POSIX`, which should be ignored.
+The [`LANG` environment variable](https://www.gnu.org/software/gettext/manual/html_node/Locale-Environment-Variables.html) specifies the user-preferred locale (in the form `ll[_CC][.encoding]`). The [`LANGUAGE` environment variable](https://www.gnu.org/software/gettext/manual/html_node/The-LANGUAGE-variable.html) specifies a priority list of locales (in the form `l1:l2:...`) that can be used if the locale defined by `LANG` is not available. Both `LANG` and `LANGUAGE` may contain the values `C` or `POSIX`, which should be ignored.
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:171]: Why not apply the same change as you did below? Even if just for consistency.


```suggestion
The [`LANG` environment variable](https://www.gnu.org/software/gettext/manual/html_node/Locale-Environment-Variables.html) specifies the user's preferred locale (in the form `ll[_CC][.encoding]`). The [`LANGUAGE` environment variable](https://www.gnu.org/software/gettext/manual/html_node/The-LANGUAGE-variable.html) specifies a priority list of locales (in the form `l1:l2:...`) that can be used if the locale defined by `LANG` is not available. Both `LANG` and `LANGUAGE` may contain the values `C` or `POSIX`, which should be ignored.
```

File: CLIENT-SPECIFICATION.md [hunk 8]
```diff
@@ -240,11 +241,11 @@ the commit hash changes when merging with squash or rebase.
 
  - [v1.3, June 11th 2020](https://github.com/tldr-pages/tldr/blob/v1.3/CLIENT-SPECIFICATION.md) ([#4101](https://github.com/tldr-pages/tldr/pull/4101))
    - Clarified fallback to English in the language resolution algorithm.
-   - Update `LANG` and `LANGUAGE` environment variable to conform to the GNU spec.
+   - Update the `LANG` and `LANGUAGE` environment variables to conform to the GNU spec.
 
  - [v1.2, July 3rd 2019](https://github.com/tldr-pages/tldr/blob/v1.2/CLIENT-SPECIFICATION.md) ([#3168](https://github.com/tldr-pages/tldr/pull/3168))
    - Addition of a new `-L, --language` recommended command-line option.
-   - Rewording of the language section also encouraging the use of configuration files for language.
+   - Rewording of the language section also encourages the use of configuration files for language.
```
Comment (waldyrious) [inline CLIENT-SPECIFICATION.md:248]: I don't think this matches the original intent of the sentence. Maybe a comma helps here?


```suggestion
   - Rewording of the language section, also encouraging the use of configuration files for language.
```

**Review by waldyrious — COMMENTED** [review]
Nice work, @kbdharun! I've left some inline suggestions.

File: CLIENT-SPECIFICATION.md (hunk 2 above)
Comment (kbdharun) [reply CLIENT-SPECIFICATION.md:5]:
```suggestion
This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. For a list of previous versions of the specification, see the [changelog section](#Changelog) below.
```

File: CLIENT-SPECIFICATION.md (hunk 4 above)
Comment (kbdharun) [reply CLIENT-SPECIFICATION.md:40]:
```suggestion
`-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if caching is supported.
```

File: CLIENT-SPECIFICATION.md [hunk 9]
```diff
@@ -116,7 +116,7 @@ The structure inside these translation folders is identical to that of the main
 
 ## Page structure
 
-Although this specification is about the interface that clients must provide, it is also worth noting that pages are written in standard [CommonMark](https://commonmark.org/), which the exception of the non-standard `{{` and `}}` syntax, which surrounds values in an example that users may edit. Clients MUST NOT break if the page format is changed within the _CommonMark_ specification.
+Although this specification is about the interface that clients must provide, it is also worth noting that pages are written in the standard [CommonMark](https://commonmark.org/), which the exception of the non-standard `{{` and `}}` syntax, which surrounds values in an example that users may edit. Clients MUST NOT break if the page format is changed within the _CommonMark_ specification.
```
[the remark this answers is not in the document]
Comment (kbdharun) [reply CLIENT-SPECIFICATION.md:119]:
```suggestion
Although this specification is about the interface that clients must provide, it is also worth noting that pages are written in standard [CommonMark](https://commonmark.org/), which the exception of the non-standard `{{` and `}}` syntax, which surrounds values in an example that users may edit. Clients MUST NOT break if the page format is changed within the _CommonMark_ specification.
```

File: CLIENT-SPECIFICATION.md (hunk 6 above)
Comment (kbdharun) [reply CLIENT-SPECIFICATION.md:154]:
```suggestion
If a page cannot be found in _any_ platform, then it is RECOMMENDED that clients display an error message with a link to create a new issue against the `tldr-pages/tldr` GitHub repository. Said link might take the following form:
```

**Review by SethFalco — APPROVED** [review]
I much prefer the new version. It's true that some changes are subjective, but it makes the document much more accessible imo. It might've been a tad too formal before.

Now was a great time to review this while dropping the `master` branch anyway, and you did a good job. :+1:

**Review by waldyrious — APPROVED** [review]
Thanks for doing this, @kbdharun — LGTM!

**Comment by acuteenvy** [comment]
If we're bumping the major version, maybe we could fix #10040 while we're at it?
It's been laying around for quite a while now, [this](https://github.com/tldr-pages/tldr/pull/5675#discussion_r609708510) is the earliest mention I could find.

**Comment by kbdharun** [comment]
> If we're bumping the major version, maybe we could fix #10040 while we're at it? It's been laying around for quite a while now, [this](https://github.com/tldr-pages/tldr/pull/5675#discussion_r609708510) is the earliest mention I could find.

Anybody working on this?

**Comment by MrNossiom** [comment]
> If we're bumping the major version, maybe we could fix #10040 while we're at it? It's been laying around for quite a while now, [this](https://github.com/tldr-pages/tldr/pull/5675#discussion_r609708510) is the earliest mention I could find.

I don't know if this spec is the place to define the tldr pages syntax. But if so, it should be more clear that is the intent and not just an aside note. This is why I suggest restructuring the `Page structure` chapter of the spec, with something like the following:

The `tldr` examples pages structure are written in the standard [CommonMark](https://commonmark.org/). However, it needed the addition of a syntax to show values the user may edit to adapt the example to its needs. These parts are indicated by surrounding double parentheses, `{{` for opening and `}}` for closing. These parentheses might need to be escaped and can be done with a backslash (`\`) right before the first opening tag. Clients MUST NOT break if the page format is changed within the CommonMark specification.

Examples :
- `ping {{example.com}}` MUST BE rendered as "ping [example.com](#)"
- `docker inspect --format="\{{range.NetworkSettings.Networks}}\{{.IPAddress}}\{{end}}" {{container}}` MUST BE rendered as "docker inspect --format='{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' [container](#)"

---

I've removed the notice at the beginning
> Although this specification is about the interface that clients must provide, [...]
It seems out of place. Either, we can create another spec for the format the `tldr` pages MUST follow and point both this spec and the contributor guide to it, either we define the format here but assume it.

I'm not sure that we need the examples if this is clear enough, you can choose whether to include them.

In any case, I think that having a place where the format is clearly defined would be nice. I use `tldr` on a daily basis, and I'm very glad this exists. Thanks

**Comment by kbdharun** [comment]
Marking the PR as a draft, later today I plan on updating it to highlight #10555

**Comment by MrNossiom** [comment]
Has changing the syntax for escaping highlight expressions been considered ?

**Comment by kbdharun** [comment]
> Has changing the syntax for escaping highlight expressions been considered ?

Currently no, but we will look into it (and other related issues) in the future.

**Offtopic:**

_Since this PR has been open for a while now (with new changes like translation archives), I will request final reviews in the chatroom and once the green light is given, I will create a release soon._
